sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -340,7 +347,12 @@ private void createRunnerAndConsumersForPTransformRecursively(
               "Unable to find active bundle for instruction id %s.",
               request.getProcessBundleSplit().getInstructionId()));
     }
-    throw new UnsupportedOperationException("TODO: BEAM-3836, support splitting within SDK.");
+    BeamFnApi.ProcessBundleSplitResponse.Builder response =
+        BeamFnApi.ProcessBundleSplitResponse.newBuilder();
+    for (BeamFnDataReadRunner channelRoot : bundleProcessor.getChannelRoots()) {
+      channelRoot.trySplit(request.getProcessBundleSplit(), response);
```

Comment:
Is it intended that all roots process the same split request?

Reply:
Yes, each root gets the full split request and extracts out the desiredsplit information for itself if there is any specific to it.